### ISSUE TITLE
tracker: `generate` composability refactor 

### ISSUE THREAD

**Issue by gante** [huggingface/transformers#30810]
## `generate` + composability = more use cases with minimal rewrites

As I write this issue, `generate` is mostly a sequential monolith. Many internal functions were added over the last two years, but navigating there as a beginner is still very tricky. It is also very challenging to adapt `generate` to different tasks and/or modalities, forcing us to overwrite the entire generate function (e.g. [RAG](https://github.com/huggingface/transformers/blob/ccdabc5642bf84849af93f591e207dc625c8e1e1/src/transformers/models/rag/modeling_rag.py#L907), [MusicGen](https://github.com/huggingface/transformers/blob/ccdabc5642bf84849af93f591e207dc625c8e1e1/src/transformers/models/musicgen/modeling_musicgen.py#L1542)). All these aspects make using, maintaining, and testing `generate` harder than it should.

This issue is a tracker for the refactor of `generate`, where we aim to build the structure outlined in [this board](https://miro.com/app/board/uXjVNgMGfaQ=/). Key ideas for this refactor:
👉 All models can use the base `generate` API
👉 Reduce if/else blocks
👉 Reduce the barriers to entry for new decoding methods/modalities/use cases
👉 Reduce per-model overwrites when possible
👉 Add unit tests
👉 Add documentation regarding the structure of `generate`

### Tasks
TBD

**Comment by amyeroberts** [comment]
Adding the WIP label, so you don't get pinged by the stale bot 🤖

**Comment by dmarx** [comment]
Could you elaborate on the "prefill" component? My impression is that this step converts the prompt into a KV cache, i.e. "pre-filling" the KV component for the tokens that are fixed. If that's correct, this component could probably serve double duty (as an exit point from the `generate` procedure) for users who just want logprobs/scores for the prompt.

**Comment by dmarx** [comment]
Upsteam of the "generate outputs" step in the decoding loop should be a templated `_prepare_outputs` function whose job is just to attach output attributes to the appropriate output class. Design POC via #29545 

* a simple _prepare_outputs function which encapsulates conditional statements that are repeated at the ends of the sampling procedures: https://github.com/huggingface/transformers/pull/29545/files#diff-26783ca033d92b4ce2e01eb691dbf5b05dd972b0cbfdc69fc726cf77a9dcb011R354-R379
* illustration of how utilizing this function simplifies the code and abstracts away logic most users won't want to be concerned with: https://github.com/huggingface/transformers/pull/29545/files#diff-26783ca033d92b4ce2e01eb691dbf5b05dd972b0cbfdc69fc726cf77a9dcb011L2201-R2267

**Comment by gante** [comment]
@dmarx 

> Could you elaborate on the "prefill" component? My impression is that this step converts the prompt into a KV cache, i.e. "pre-filling" the KV component for the tokens that are fixed. If that's correct, this component could probably serve double duty (as an exit point from the generate procedure) for users who just want logprobs/scores for the prompt.

Correct, it can be made a public function with that additional purpose. The difference between the prefill for `generate` and obtaining the scores for the prompt is that in the former, we only want to keep the past KV. The different output needs suggest to me that a stand-alone public function is preferable to an alternate exit to `generate` :D 

> IMHO, upsteam of (part of?) the "generate outputs" step in the decoding loop should be a templated _prepare_outputs function whose job is just to attach output attributes to the appropriate output class. Design POC via https://github.com/huggingface/transformers/pull/29545, concretely:
https://github.com/coreweave/transformers/blob/dmarx.output_streamer/src/transformers/generation/utils.py#L354-L378

That's a good idea, adding it to the diagram!

**Comment by jiqing-feng** [comment]
Hi @gante . I see this issue has been open for 2 months. Do you have any plans to start the 1st step?  If it relies on other tasks and cannot start recently, do you mind considering #31564 as the temporary solution?

**Comment by dmarx** [comment]
@gante maybe you could create sub-issues to track progress on dependent sub tasks and centralize visibility for any in-progress work?

**Comment by gante** [comment]
@tylerweitzman after this plan is complete, streaming should yield all model outputs (and not just the tokens).

The tricky part is timeline, as this competes with other projects 🤗 I'm currently working on #28981 and #32685, which have a higher priority

**Comment by dmarx** [comment]
@tylerweitzman workaround available here: https://github.com/huggingface/transformers/pull/29545

**Comment by qgallouedec** [comment]
Closing this one. gante has left HF, and this tracker is his personal roadmap, so it no longer has an owner. Nothing has referenced it since 2024, including the PRs that actually completed task 1.

Where the plan stands on main:

- Task 1 is done. `_prefill` lives in `generation/utils.py` and is used by `_sample` and `_beam_search`, landed in #40652 and refined in #42088. The final design differs from the description above: we prefill the full prompt and skip the first forward inside the loop, rather than excluding the last token.
- Task 6's `LogitsWarper` deprecation is done, the base class is gone.
- Tasks 2, 4, 5, 7 were never started. `generate` is still one 420-line function.
- Task 3 is not happening in this form. Streaming is covered by `TextIteratorStreamer`, `AsyncTextIteratorStreamer`, `transformers serve` and `transformers chat`. 3e was closed as not planned in #30487.
- Tasks 9 to 12 were never started. 45 model files still define their own `generate`, rag and musicgen included.

Generate did get simpler over the last two years, just not through this plan: decoding methods went from 8 to 3 (`_sample`, `_beam_search`, `_assisted_decoding`) as v5 dropped contrastive search, beam sample, group beam search and constrained beam search, and the composability effort went into continuous batching and `transformers serve` instead.

@jiqing-feng task 1 shipped in #40652, apologies for the very late follow up.

@tylerweitzman streaming full model outputs is still unsupported. #29545 was closed without merging, so that is a fork patch rather than something you can get from transformers. If you still need it, open a dedicated feature request and we can look at it on its own merits.

If someone on the generation side wants to keep part of this, please open a fresh tracker scoped to what is on the roadmap today. The board is still readable
